### PR TITLE
Allow enabling/disabling certain components

### PR DESCRIPTION
**Pull request by jjwilke**
Draft of proposal for enabling/disabling particular components

### CODE REVIEW

**Comment by jjwilke** [comment]
Main design question is how much granularity we want in turning things off/on.  Right now we have a single granularity: all of blas, sparse, batched, or graph.

File: perf_test/CMakeLists.txt
Comment (lucbv) [inline perf_test/CMakeLists.txt:16]: Same as above, I would add a "Common" component here

File: perf_test/CMakeLists.txt
Comment (lucbv) [inline perf_test/CMakeLists.txt:18]: I am OK with merging all the Blas into a single component.

File: unit_test/CMakeLists.txt
Comment (lucbv) [inline unit_test/CMakeLists.txt:42]: I see from above that you refactored the code a bit to only use the new `KOKKOSKERNELS_ADD_UNIT_TEST` and now we add a new field `COMPONENT` to each test so that they get compiled only `KOKKOSKERNELS_IS_ENABLED` returns true for that particular component.

**Review by lucbv — COMMENTED** [review]
@jjwilke only a few comments, mainly wondering if we can have a `Common` component as we have it as one of our unit-test.
Also I think we could expand the `component` approach to also cover the examples but that can definitely wait for another PR!
Overall I am happy with this : )

**Comment by lucbv** [comment]
@brian-kelley @ndellingwood any opinion on this?
It will be useful for me to get this merged to continue some work with HIP

**Review by ndellingwood — APPROVED** [review]
LGTM from a scan of the changes, this is cool thanks @jjwilke 
Approved though we need to run tests and should test configurations with Trilinos also before merging

**Comment by ndellingwood** [comment]
@lucbv have you run any tests with the changes yet? I can do some Trilinos configuration testing tomorrow

**Comment by lucbv** [comment]
Actually I have not but I can run them tonight, assuming that the default `ALL` value for `COMPONENTS` works fine I think this should not cause too much issues with Trilinos but it's worth checking.

**Review by brian-kelley — COMMENTED** [review]
Everything looks good. The one thing I'm slightly concerned about is what happens if only sparse is enabled, since some sparse features also use graph and blas features. Do the graph/blas headers still get installed if those components aren't enabled? If graph/blas aren't ETI'd, can things like gauss-seidel use them?

This would also be a consideration if you added a common component, because pretty much everything depends on those headers (although nothing in common is ETI'd)

**Review by brian-kelley — APPROVED** [review]
Actually, now that I look again it doesn't seem like the components change which headers get installed.

**Comment by ndellingwood** [comment]
Trilinos fails to configure with the PR
```bash
CMake Error at kokkos-kernels/cmake/kokkoskernels_tribits.cmake:135 (VERIFY_EMPTY):
  Unknown CMake command "VERIFY_EMPTY".
```

Change to get things configuring is pretty simple, move the `VERIFY_EMPTY` function from `fake_tribits.cmake` to `kokkoskernels_tribits.cmake`. I verified this resolved the Trilinos configuration issue.

**Comment by ndellingwood** [comment]
Started spot-check with https://github.com/jjwilke/kokkos-kernels/pull/1

**Comment by lucbv** [comment]
@jjwilke @ndellingwood 
Where are we on this PR at the moment?

**Comment by ndellingwood** [comment]
kokkos-dev-2 spot-check (no cmake config options to disable components, hitting the "ALL" case):

```bash
Running on machine: kokkos-dev-2

KokkosKernels Repository Status:  d079c22da6362217052a629ef4bc7753c9700ddf Merge pull request #1 from ndellingwood/enable-components

Kokkos Repository Status:  9668221ee442cce8ec3e9d2f61d14b52bfd06cb3 Merge pull request #3491 from dalg24/ompt_team_policy_missing_converting_constructor

#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=662 run_time=124
clang-8.0-Pthread_Serial-release build_time=220 run_time=115
clang-9.0.0-Pthread-release build_time=130 run_time=54
clang-9.0.0-Serial-release build_time=132 run_time=45
cuda-10.1-Cuda_OpenMP-release build_time=889 run_time=124
cuda-9.2-Cuda_Serial-release build_time=915 run_time=186
gcc-7.3.0-OpenMP-release build_time=149 run_time=44
gcc-7.3.0-Pthread-release build_time=124 run_time=56
gcc-8.3.0-Serial-release build_time=151 run_time=47
gcc-9.1-OpenMP-release build_time=184 run_time=44
gcc-9.1-Serial-release build_time=165 run_time=46
intel-17.0.1-Serial-release build_time=291 run_time=48
intel-18.0.5-OpenMP-release build_time=400 run_time=47
intel-19.0.5-Pthread-release build_time=397 run_time=56
```

I previously confirmed the changes configure with Trilinos

**Comment by lucbv** [comment]
@ndellingwood thanks for confirming, I guess we can merge this then unless someone has an objection?
This would be nice for me as I plan to rebase my HIP work on it : )

**Comment by lucbv** [comment]
Thanks @jjwilke and @ndellingwood for helping with this, it should pay off for more than just the HIP backend : )

**Pull request merged**
